### ISSUE TITLE
questions about pretrained vgg16 backbone for training faster rcnn?

### ISSUE THREAD

**Issue by cizhenshi** [open-mmlab/mmdetection#1808]
I am using vgg16-faster RCNN to train cityscapes. I find that the model trained with mmdetection vgg16 backbone is much lower than the model trained with other faster RCNN implementations (I use https://github.com/jwyang/faster-rcnn.pytorch)  (map 47 for faster-rcnn. pytorch and map 34 for mmdetection). My config file is as follows:

# model settings
model = dict(
    type='FasterRCNN',
    # pretrained='open-mmlab://vgg16_caffe',
    backbone=dict(
        type='VGG',
        depth=16,
        num_stages=5,
        out_indices=(4,),
        frozen_stages=2,
        style='pytorch'),
    rpn_head=dict(
        type='RPNHead',
        in_channels=512,
        feat_channels=512,
        anchor_scales=[8, 16, 32],
        anchor_ratios=[0.5, 1.0, 2.0],
        anchor_strides=[16],
        target_means=[.0, .0, .0, .0],
        target_stds=[1.0, 1.0, 1.0, 1.0],
        loss_cls=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
        loss_bbox=dict(type='SmoothL1Loss', beta=1.0 / 9.0, loss_weight=1.0)),
    bbox_roi_extractor=dict(
        type='SingleRoIExtractor',
        roi_layer=dict(type='RoIAlign', out_size=7, sample_num=2),
        out_channels=512,
        featmap_strides=[16]),
    bbox_head=dict(
            type='SharedFCBBoxHead',
            num_fcs=2,
            in_channels=512,
            fc_out_channels=4096,
            roi_feat_size=7,
            num_classes=9,
            target_means=[0., 0., 0., 0.],
            target_stds=[0.1, 0.1, 0.2, 0.2],
            reg_class_agnostic=False,
            loss_cls=dict(
                type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
            loss_bbox=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.0)),)
# model training and testing settings
train_cfg = dict(
    rpn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.7,
            neg_iou_thr=0.3,
            min_pos_iou=0.3,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=256,
            pos_fraction=0.5,
            neg_pos_ub=-1,
            add_gt_as_proposals=False),
        allowed_border=0,
        pos_weight=-1,
        debug=False),
    rpn_proposal=dict(
        nms_across_levels=False,
        nms_pre=12000,
        nms_post=2000,
        max_num=2000,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.5,
            min_pos_iou=0.5,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=256,
            pos_fraction=0.25,
            neg_pos_ub=-1,
            add_gt_as_proposals=True),
        pos_weight=-1,
        debug=False))
test_cfg = dict(
    rpn=dict(
        nms_across_levels=False,
        nms_pre=6000,
        nms_post=300,
        max_num=300,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=dict(
        score_thr=0.05, nms=dict(type='nms', iou_thr=0.5), max_per_img=300))
# dataset settings
dataset_type = 'CityscapesDataset'
data_root = '../data/cityscapes/'
img_norm_cfg = dict(mean=[102.9801, 115.9465, 122.7717], std=[1, 1, 1], to_rgb=False)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', img_scale=(1200, 600), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1200, 600),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='RandomFlip'),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]
data = dict(
    imgs_per_gpu=2,
    workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        ann_file=data_root + 'cityscapes_train.json',
        img_prefix=data_root,
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        ann_file=data_root + 'cityscapes_val.json',
        img_prefix=data_root,
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=data_root + 'cityscapes_val.json',
        img_prefix=data_root,
        pipeline=test_pipeline))
# optimizer
optimizer = dict(type='SGD', lr=5e-3, momentum=0.9, weight_decay=0.0005)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
# learning policy
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    step=[8, 11])
checkpoint_config = dict(interval=1)
# yapf:disable
log_config = dict(
    interval=50,
    hooks=[
        dict(type='TextLoggerHook'),
        dict(type='TensorboardLoggerHook')
    ])
# yapf:enable
# runtime settings
total_epochs = 12
dist_params = dict(backend='nccl')
log_level = 'INFO'
work_dir = './work_dirs/city_base'
load_from = "../data/vgg16_caffe.pth"
resume_from = None
workflow = [('train', 1)]

the checkpoint load from is converted from vgg16_caffe pretrained models, so the mean and std is different from mmdetection.

Can you give me some helps?

**Comment by cizhenshi** [comment]
faster-rcnn.pytorch result:
faster rcnn.pytorch  检测结果

AP for person = 0.4058
AP for rider = 0.4995
AP for car = 0.5373
AP for truck = 0.3605
AP for bus = 0.6009
AP for train = 0.5106
AP for motorcycle = 0.4145
AP for bicycle = 0.4270
Mean AP = 0.4695

mmdetection result:
+------------+------+------+--------+-----------+-------+
| class      | gts  | dets | recall | precision | ap    |
+------------+------+------+--------+-----------+-------+
| person     | 3419 | 1415 | 0.219  | 0.530     | 0.205 |
| rider      | 556  | 364  | 0.326  | 0.497     | 0.300 |
| car        | 4667 | 2551 | 0.367  | 0.671     | 0.360 |
| truck      | 93   | 217  | 0.441  | 0.189     | 0.355 |
| bus        | 98   | 198  | 0.622  | 0.308     | 0.550 |
| train      | 23   | 101  | 0.565  | 0.129     | 0.483 |
| motorcycle | 149  | 173  | 0.336  | 0.289     | 0.289 |
| bicycle    | 1175 | 855  | 0.317  | 0.436     | 0.271 |
+------------+------+------+--------+-----------+-------+
| mAP        |      |      |        |           | 0.352 |
+------------+------+------+--------+-----------+-------+

**Comment by hellock** [comment]
Are the model architecture and training settings the same?

**Comment by cizhenshi** [comment]
After my modification, the architecture of the model is basically the same, except that my mmdetection implementation does not use dropout in the Shared fc， but faster rcnn.pytorch used. I visualized the results of some images and found that many of the small objects were not detected.

**Comment by cizhenshi** [comment]
For faster-rcnn.pytorch, it first flipped all the images offline horizontally, and then formed a training set with the original images to train 9 epochs, and conducted a learning rate decrease at the 8th epoch. For mmdetection, I tried 1x and 2x with random hflip and got approximate results, 0.34+ map.

**Comment by cizhenshi** [comment]
I modified the anchor scales to [2,4,8,16,32] instead of [8,16,32], and I got map of 50.6. I noticed that resnet50-c4 also set 5 anchors in the same way. Is there a deviation in my understanding of the anchor of mmdetection? For Faster RCNN, we should choose three scales and three ratios, is right?

**Comment by cizhenshi** [comment]
I don't know whether the definition of anchor in mmdetection is exactly the same as py-faster-rcnn, but after I changed the anchor scales to [4,8,16], I got mAP 47.9.This result looks similar to faster rcnn.pytorch. Faster rcnn.pytorch used the same configuration as py-faster-rcnn, with ANCHOR_SCALES = [8,16,32].  I have reading some code about anchor generator and don't have find something different.  So I don't know why there's this performance gap.  Can you give me some tips?

**Comment by hellock** [comment]
1. You can also try the same lr schedule as in faster-rcnn.pytorch, which corresponds to 18 epochs in total and a decrease in the 16th epoch.
2. For the anchor settings, we just follow the typical convention and also the official release of detectron. On  COCO dataset, C4 models adopts 5 scales and FPN models adopts 3 scales for each level.
3. I have not read the code of faster-rcnn.pytorch  so that I cannot conclude whether the anchors are the same. You may look into the implementation details of it.

**Comment by cizhenshi** [comment]
Thank you for your answer. After looking into the implementation of mmdetection and faster-rcnn.pytorch, I think the problem may be on the set of **min_pos_iou**. when the setting of anchors does not accord with object distribution, min_pos_iou = 0.5 may cause a lot of performance degradation, especially in the absence of FPN structure. I have set min_pos_iou to  0 and  map from 0.34 to  0.41 in the case of 1 x.  I will continue to try 2 x and report on the issue.

**Comment by freepoet** [comment]
Hi,bro!~</br>
I want to train faster rcnn with pretrained vgg16 backbone, could you provide your code for reference?</br>
thx!
